hamilton/lifecycle/api.py

```diff
@@ -371,10 +373,121 @@ def run_after_graph_execution(
         pass
 
 
+class TaskSubmissionHook(BasePreTaskSubmission, abc.ABC):
+    """Implement this to hook into the task submission process. Tasks are submitted to an executor,
+    which then controls how and where the nodes associated with the task are run."""
+
+    @override
+    def pre_task_submission(
+        self,
+        *,
+        run_id: str,
+        task_id: str,
+        nodes: List["node.Node"],
+        inputs: Dict[str, Any],
+        overrides: Dict[str, Any],
+        spawning_task_id: Optional[str],
+        purpose: NodeGroupPurpose,
+    ):
+        self.run_before_task_submission(
+            run_id=run_id,
+            task_id=task_id,
+            nodes=nodes,
+            inputs=inputs,
+            overrides=overrides,
+            spawning_task_id=spawning_task_id,
+            purpose=purpose,
+        )
+
+    @abc.abstractmethod
+    def run_before_task_submission(
+        self,
+        *,
+        run_id: str,
+        task_id: str,
+        nodes: List["node.Node"],
+        inputs: Dict[str, Any],
+        overrides: Dict[str, Any],
+        spawning_task_id: Optional[str],
+        purpose: NodeGroupPurpose,
+        **future_kwargs,
+    ):
+        """Runs prior to a task being submitted to an executor. By definition this is run *outside*
+        of the task executor, on the process that executed the driver.
+
+        :param run_id: ID of the run this is under.
+        :param task_id: ID of the task we're launching.
+        :param nodes: Nodes that are part of this task
+        :param inputs: Inputs to the task
+        :param overrides: Overrides passed to the task
+        :param spawning_task_id: ID of the task that spawned this task
+        :param purpose: Purpose of the current task group
+        :param future_kwargs: Reserved for backwards compatibility.
+        """
+        pass
+
+
+class TaskResolutionHook(BasePostTaskResolution, abc.ABC):
+    """Implement this to hook into the task resolution process. Tasks are submitted to an executor,
```

Comment:
Let's clarify a little more what task "resolution" is for users -- it's nuanced and a bit confusing.

IIUC this is what happens when we have a task "template", which gets realized/ "bound" to tasks. E.G. when all the dependencies are satisfied and we can enqueue it. Right?

Reply:
I struggled with the name of this hook/adapter. Currently the `post_task_resolution` hook gets called in `run_graph_to_completion` after the task future gets resolved. Maybe it's better to describe it as 'a hook that fires after a task returns from an executor'? Ha - that is actually a better complement to `pre_task_submission`! What do you think - I would happy to rename the hook/adapter - maybe `BasePostTaskReturn` / `TaskReturnnHook`?

Reply:
Yeah I like `post_task_return` and `TaskReturnHooks` -- I think that's a bit clearer.